### ISSUE TITLE
Is passing objects as reference from a coroutine to a coroutine function ok?

### ISSUE THREAD

**Issue by BurievSardor** [danvratil/qcoro#131]
An example:

QCoro::Task<> MyClass::doSomethingWithData(QByteArray &data) {

	...........................................

	QNetworkAccessManager networkAccessManager;
	const QNetworkReply *reply = co_await networkAccessManager.get(url);
	...............................................
}

QCoro::Task<> MyClass::fetchData() {

	QNetworkAccessManager networkAccessManager;
	const QNetworkReply *reply = co_await networkAccessManager.get(url);
	const auto data = reply->readAll();
	doSomethingWithData(data)
}

**Comment by danvratil** [comment]
Generally speaking, yes. It is OK to use references as coroutine arguments, but one has to be careful about the lifetime of the referenced object.

The example above may or may not be safe, since you did not show how you are using the `data` reference in `doSomethingWithData()`.

Let me expand on your example:
```cpp
QCoro::Task<> MyClass::doSomethingWithData(QByteArray &data) {
    QNetworkAccessManager networkAccessManager;
    // It's perfectly safe to use `data` up to this point
    // Actually, it's safe to use is on the next line as well, but only in the
    // expression *after* the `co_await` keyword:
    //                                    v---------------------------v
    const QNetworkReply *reply = co_await networkAccessManager.get(url);    
    // It's *NOT* safe to use `data` from this point on. The reference is dangling.
}
```

The reason is that the `doSomethingWithData()` coroutine is *not* `co_await`ed inside the `fetchData()`  coroutine, so the following things happen:
1) the `data` variable is allocated and initialized with result of `reply->readAll()`
1) the `doSomethingWithData()` coroutine is called, passing it reference to `data`
1) at this point, accessing the `data` reference inside `doSomethingWithData()` coroutine is completely safe
1) the `doSomethingWithData()` coroutine is suspended when `co_await`ing the network reply
1) execution returns back to `fetchData()`
1) `fetchData()` coroutine reaches the end of the function body, `data` is destroyed, `networkAccessManager` is destroyed, `reply` leaks ;-)
1) at some point later, `doSomethingWithData()` coroutine is resumed
1) at this point, `data` inside `doSomethingWithData()` coroutine is a dangling reference and must NOT be used, because the object that it references has been already destroyed in step 6.

_IF_ the `doSomethingWithData()` coroutine were `co_await`ed inside `fetchData()`, then it would be totally safe to use the `data` reference anywhere inside the `doSomethingWithData()` coroutine, since the `fetchData()` coroutine would not be finished until the `doSomethingWithData()` coroutine would finish, and thus it would be guaranteed that the `data` object outlives the `data` reference.

To sum it up, yes, it's safe to pass arguments to coroutines as references, but one must be careful when using references whose lifespan crosses a suspension point (a `co_await`), as the may become dangling references while the coroutine is suspended.

**Comment by danvratil** [comment]
I should document this better in the docs :-)
